dev/dfi/packet.c

```diff
@@ -161,6 +412,229 @@ void dfi_rx_packet_buffer_data_compare(dfi_rx_packet_buffer_t *buffer,
+    offset_remainder = ((dram_cfg->t_sh_train + dram_cfg->phy_wr_en) & ratio);
+
+    // Add one extra DFI cycle if tDHtrain is odd value
+    if (offset_remainder > 0)
+    {
+        time_offset++;
+    }
+    create_ck_packet_sequence(buffer, time_offset);
+
+    // Generate CA data packets to send training sequence
+    time_offset = buffer->ts_last_packet + ((dram_cfg->t_vref_ca_long - dram_cfg->t_sh_train) >> ratio);
+    offset_remainder = (dram_cfg->t_vref_ca_long & ratio);
+    // Add one extra DFI cycle if tDHtrain is odd value
+    if (offset_remainder > 0)
+    {
+        time_offset++;
+    }
+
+    create_packet(buffer, &packet);
+    if (packet == NULL)
+    {
+        return WDDR_ERROR;
+    }
+
+    packet->packet.packet.dce_p0 = 1;
+    packet->packet.packet.dce_p1 = 1;
+    packet->packet.packet.dce_p2 = 1;
+    packet->packet.packet.dce_p3 = 1;
+
+    // Command
+    packet->packet.packet.cs_p0 = CS_HIGH << cs;
+    packet->packet.packet.cs_p1 = CS_HIGH << cs;
+    packet->packet.packet.address_p0 = command_address;
+
+    // Add to list
+    vListInitialiseItem(&packet->list_item);
+    packet->packet.packet.time = time_offset;
+    listSET_LIST_ITEM_VALUE(&packet->list_item, time_offset);
+    listSET_LIST_ITEM_OWNER(&packet->list_item, packet);
+    vListInsertEnd(&buffer->list, &packet->list_item);
+
+    // Set buffer last packet timestamp before calling create CK
+    buffer->ts_last_packet = time_offset;
+    time_offset = 1;
+
+    create_ck_packet_sequence(buffer, time_offset);
+
+    create_ck_packet_sequence(buffer, 2);
+
+    return WDDR_SUCCESS;
+}
+
+wddr_return_t create_wrlvl_packet_sequence(dfi_tx_packet_buffer_t *buffer,
+                                        wddr_freq_ratio_t ratio,
+                                        chipselect_t cs,
+                                        uint16_t time_offset)
+{
+    packet_item_t *packet;
+    // Need at least 2 cycles of DQS to send
+    uint8_t num_packets = (2 >> ratio);
+
+    // Initialize time_offset with last packet timestamp
+    time_offset += buffer->ts_last_packet;
+
+    for (uint8_t i = 0; i < num_packets; i++)
+    {
+        create_packet(buffer, &packet);
+        if (packet == NULL)
+        {
+            return WDDR_ERROR;
+        }
+
+        packet->packet.packet.dce_p0 = 1;
+        packet->packet.packet.dce_p1 = 1;
+        packet->packet.packet.dce_p2 = 1;
+        packet->packet.packet.dce_p3 = 1;
+        packet->packet.packet.cke_p0 = DFI_PACK_CKE_VAL;
+        packet->packet.packet.cke_p1 = DFI_PACK_CKE_VAL;
+        packet->packet.packet.cke_p2 = DFI_PACK_CKE_VAL;
+        packet->packet.packet.cke_p3 = DFI_PACK_CKE_VAL;
+
+        packet->packet.packet.wrdata_en_p0 = 1;
+        packet->packet.packet.wrdata_en_p1 = 1;
+
+        if (ratio == WDDR_FREQ_RATIO_1TO2)
+        {
+            packet->packet.packet.wrdata_en_p2 = 1;
+            packet->packet.packet.wrdata_en_p3 = 1;
+        }
```

Comment:
Do we need this if statement?

Reply:
This no long applies now that we use the fill_wrdata_en_packet api function